processor/lsmintervalprocessor/internal/merger/limits/tracker.go

```diff
@@ -0,0 +1,297 @@
+// under the License.
+
+package limits // import "github.com/elastic/opentelemetry-collector-components/processor/lsmintervalprocessor/internal/merger/limits"
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"hash"
+	"slices"
+
+	"github.com/axiomhq/hyperloglog"
+)
+
+// Tracker tracks the configured limits while merging. It records the
+// observed count as well as the unique overflow counts.
+type Tracker struct {
+	maxCardinality uint64
+	// Note that overflow buckets will NOT be counted in observed count
+	// though, overflow buckets can have overflow of their own.
+	observedCount  uint64
+	overflowCounts *hyperloglog.Sketch
+}
+
+func newTracker(maxCardinality uint64) *Tracker {
+	return &Tracker{maxCardinality: maxCardinality}
+}
+
+func (t *Tracker) Equal(other *Tracker) bool {
+	if t.maxCardinality != other.maxCardinality {
+		return false
+	}
+	if t.observedCount != other.observedCount {
+		return false
+	}
+	return t.EstimateOverflow() == other.EstimateOverflow()
+}
+
+func (t *Tracker) HasOverflow() bool {
+	return t.overflowCounts != nil
+}
+
+func (t *Tracker) EstimateOverflow() uint64 {
+	if t.overflowCounts == nil {
+		return 0
+	}
+	return t.overflowCounts.Estimate()
+}
+
+// CheckOverflow checks if overflow will happen on addition of a new entry with
+// the provided hash denoting the entries ID. It assumes that any entry passed
+// to this method is a NEW entry and the check for this is left to the caller.
+func (t *Tracker) CheckOverflow(f func() hash.Hash64) bool {
+	if t.maxCardinality == 0 {
+		return false
+	}
+	if t.observedCount == t.maxCardinality {
+		if t.overflowCounts == nil {
+			// Creates an overflow with 14 precision
+			t.overflowCounts = hyperloglog.New14()
+		}
+		t.overflowCounts.InsertHash(f().Sum64())
+		return true
+	}
+	t.observedCount++
+	return false
+}
+
+// MergeEstimators merges the overflow estimators for the two trackers.
+// Note that other required maintenance of the tracker for merge needs to
+// done by the caller.
+func (t *Tracker) MergeEstimators(other *Tracker) error {
+	if other.overflowCounts == nil {
+		// nothing to merge
+		return nil
+	}
+	if t.overflowCounts == nil {
+		t.overflowCounts = other.overflowCounts.Clone()
+		return nil
+	}
+	return t.overflowCounts.Merge(other.overflowCounts)
+}
+
+// Marshal marshals the tracker to a byte slice.
+func (t *Tracker) Marshal() ([]byte, error) {
+	// TODO (lahsivjar): Estimate the size of the trackers to optimize
+	// allocations. Also see: https://github.com/axiomhq/hyperloglog/issues/44
+	b := make([]byte, 8) // reserved for observed count
+
+	var offset int
+	binary.BigEndian.PutUint64(b[offset:(offset+8)], t.observedCount)
+	offset += 8
+	if t.overflowCounts != nil {
+		hll, err := t.overflowCounts.MarshalBinary()
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal limits: %w", err)
+		}
+		b = slices.Grow(b, len(hll))[:offset+len(hll)]
+		copy(b[offset:], hll)
```

Comment:
We should look at making the HLL++ implement the coming-soon BinaryAppender interface: https://pkg.go.dev/encoding@master#BinaryAppender

I'll look into this: https://github.com/axiomhq/hyperloglog/pull/47